Update subMenuBuilder so "installer-entry" is always at the top

sebastian-marinescu opened the pull request:
Due to the default-value of menuindex in menu-entries of newly installed extras being most of the time "0", the installer entry will jump up and/or down after an extra gets installed/updated. 

Therefore it should at least be fixed at the bottom of the dropdown-menu. 

Fixes #11694 pragmatically

sebastian-marinescu committed:
Update installer-entry to always be at the bottom

Due to the default-value of menuindex in menu-entries of newly installed extras being most of the time "0", the installer entry will jump and/or down after an extra gets installed/updated. 

Therefore it should at least be fixed at the bottom of the dropdown-menu. 

Fixes #11694 pragmatically

Mark-H reviewed (changes requested):
With all due respect and appreciation for opening a PR, I don't consider putting it at the bottom an improvement. [Posted an alternative suggestion in the issue](https://github.com/modxcms/revolution/issues/11694#issuecomment-472009581).

sebastian-marinescu committed:
Update transport.core.menus.php

sebastian-marinescu committed:
"Installer" always as first child

sebastian-marinescu commented:
> With all due respect and appreciation for opening a PR, I don't consider putting it at the bottom an improvement. [Posted an alternative suggestion in the issue](https://github.com/modxcms/revolution/issues/11694#issuecomment-472009581).

Changed the behavior to suggested one.

JoshuaLuckers commented:
@sebastian-marinescu did you look if this was possible?

> What I'd suggest we try (I haven't looked at the relevant code yet) is to add some extra code into the menu builder that makes sure the installer is always first. 
> 
> Something like an `ORDER BY field(modMenu.key, 'installer') ASC` (but in xPDO) before the menuindex sort is applied.

_Originally posted by @Mark-H in https://github.com/modxcms/revolution/issues/11694#issuecomment-472009581_

Mark-H commented:
Hey @sebastian-marinescu - did you determine sorting in the SQL as I suggested wasn't possible and that the html-replacement technique is best? Just hoping for some feedback from you here as I'm not clear on what this PR is at and 2.7.2 is coming soon.

Ibochkarev commented:
Fixed #14494 

@modxbot close

Ibochkarev commented:
@modxbot open

SnowCreative commented:
Even better would be to make the installer/updater scripts NOT CHANGE the menuindex on already installed items. Every time an extra gets updated, its position in the menu changes. Whatever order has been set by users should be retained after updates. New extras could go at the bottom, and then people could drag them where they want in the menu.

The same thing goes for menu parents. For example, I use the ClientConfig extra, and it makes the most sense to me and my clients to keep that under the Site (Content) menu. Every update moves it back under the Extras menu, and I have to move it back where I want it manually.

Mark-H commented:
That's not something the core does - that's each extra saying their menu object should be updated.

Jako commented:
So it is a package build issue. And these methods are in the core.

Ibochkarev commented:
@Mark-H @Jako Guys, what in the end on this PR needs to be done?

Mark-H commented:
@Ibochkarev [In the related issue](https://github.com/modxcms/revolution/issues/11694#issuecomment-472009581) I posted an alternative solution before this PR was created that handled it in the menu generation and I've asked if @sebastian-marinescu tried that or not. If my idea, which seems like a cleaner solution, was not feasible I'd happily merge this but I'm not sure if it was tried.

sebastian-marinescu commented:
Sorry for ghosting people. 

I just wasn't sure back then where/how the menu gets grabbed. 
Trying to re-track it, I think I found it: 

- The [modManagerController](https://github.com/modxcms/revolution/blob/2.x/core/model/modx/modmanagercontroller.class.php#L178) loads the header, which is just an include of the [header.php](https://github.com/modxcms/revolution/blob/2.x/manager/controllers/default/header.php) in the default manager-theme

- Inside the header.php the [`buildMenu`](https://github.com/modxcms/revolution/blob/2.x/manager/controllers/default/header.php#L150)-method gets the [menus from cache](https://github.com/modxcms/revolution/blob/2.x/manager/controllers/default/header.php#L225)

- If it's not already cached, it [rebuilds the menu-cache](https://github.com/modxcms/revolution/blob/2.x/manager/controllers/default/header.php#L242)

- Inside the `getSubMenus`-method it [sorts by `menuindex`](https://github.com/modxcms/revolution/blob/2.x/core/model/modx/modmenu.class.php#L104)

So this is where you would add an extra sort-criteria @Mark-H ? 
```
$c->sortby($this->xpdo->getSelectColumns('modMenu','modMenu','',array('menuindex')),'ASC');
``` 
FYI: @Ibochkarev

Mark-H commented:
Yep, that's what I was hoping for. :)

sebastian-marinescu committed:
Update header.php

sebastian-marinescu committed:
Update modmenu.class.php

make sure installer is always on top

sebastian-marinescu commented:
So I think I found the solution and pushed it to my [patch-3](https://github.com/sebastian-marinescu/revolution/tree/patch-3) branch - but it doesn't show here on the PR like it used to do...

Anyone any idea how I can update the commits here?

sebastian-marinescu commented:
Oh okay, the friendly cla-bot unfroze the situation - nice :) 
Hopefully this is the right way now @Mark-H @opengeek

Jako commented:
Would it be possible to have folders first (maybe by an option)?

JoshuaLuckers commented:
I would be happy with this alone!

sebastian-marinescu commented:
> Would it be possible to have folders first (maybe by an option)?

I'm not sure I understand. Do you mean for the two entries in "Manage"? 
![image](https://user-images.githubusercontent.com/3798871/87697723-c0adf500-c792-11ea-8b12-f966a3dcf019.png)

Would probably be possible, by checking if it is a parent - but I don't want to mix this PR up: I just want the "Installer" entry to not jump around anymore, and I'm hoping to get that into the next release.

Jako commented:
I often put folders first in the Extras menu to avoid things like in this screenshot at the bottom of the browser window.

<img width="559" alt="Cut off menu" src="https://user-images.githubusercontent.com/148371/87716607-5e172200-c7af-11ea-9429-9bfd3168ffe4.png">

With your patch a submenu could start only at the second entry, which is a little less space to the bottom. it would be great, if entries with a submenu could stay in the before the Installer (by enabling a system setting).

Ruslan-Aleev commented:
> With your patch a submenu could start only at the second entry, which is a little less space to the bottom. it would be great, if entries with a submenu could stay in the before the Installer (by enabling a system setting).

Maybe we should add this fix to the 2.x branch - https://github.com/modxcms/revolution/pull/14300? And for 3.x right away.

JoshuaLuckers reviewed (approved):
I think it's a good idea to solve this little annoyance this way. In 3.x we can improve it further by implementation one of the suggestions.

sebastian-marinescu commented:
> With your patch a submenu could start only at the second entry, which is a little less space to the bottom. it would be great, if entries with a submenu could stay in the before the Installer (by enabling a system setting).

@Jako I see now, thx. So your setting should/would only be relevant to the "Extras" main-menu or to all? 
And while I understand your use-case, I'd still be in favor of having the "installer" as first, *always* - then your special-folders. It would still work. 

Another thing I wanted to point out is, that your component descriptions take a lot of space away - 
maybe you could check out "AdminTools", which hides the descriptions and therefore gives much more space: 

![image](https://user-images.githubusercontent.com/3798871/87790752-628c1b00-c841-11ea-99ce-82d45d30c39b.png)

SnowCreative commented:
I really don't understand why the consensus is that the installer should always be at the top. Once a site is set up, that is going to be the LEAST USED item in the list. Putting it at the bottom, or else just sorting it like everything else, makes a lot more sense to me.

Ruslan-Aleev commented:
> Once a site is set up, that is going to be the LEAST USED item in the list.

Personally, I am enraged that the "Installer" jumps in menu :)
And while you **set up the site once**, you get tired of looking for the installer in the menu.

SnowCreative commented:
> > Once a site is set up, that is going to be the LEAST USED item in the list.
> 
> Personally, I am enraged that the "Installer" jumps in menu :)
> And while you **set up the site once**, you get tired of looking for the installer in the menu.

If it's sorted alphabetically, you can find it. If it's always at the bottom, you can find it. When everything is jumping around, as they do now, you can't find any of the plugins, including the installer.

Ruslan-Aleev commented:
> If it's sorted alphabetically, you can find it. If it's always at the bottom, you can find it.

- Alphabetically, it is located only for the English language.
- The bottom location is not convenient, because the menu item can be clipped by the screen (@Jako pointed out about this above).
- **Positioning on top** is the most effective UX option, in my opinion.

SnowCreative commented:
> > If it's sorted alphabetically, you can find it. If it's always at the bottom, you can find it.
> 
>     * Alphabetically, it is located only for the English language.
> 
>     * The bottom location is not convenient, because the menu item can be clipped by the screen (@Jako pointed out about this above).
> 
>     * **Positioning on top** is the most effective UX option, in my opinion.
If it gets clipped by the screen at the bottom, then moving it to the top will make something else get clipped by the screen instead. That's a separate issue, which also needs to get fixed. As for alpha sorting, the sorting would have to be done depending on the language selected for the manager itself. There is a separate thread, I believe, asking for alpha sorting of plugins rather than using menuindex as is now done, and which is responsible for making the menu a mess since the menuindexes get reset with every update. Any custom permission you've put on a plugin also gets wiped out when the plugin is updated, which is yet another issue.

Why can't we just install a system setting to determine where to put the installer, instead of trying to force everyone to have it in the same place?

Or, better yet, why not just **remove the possibility for package updates to reset the menuindexes at all**. That would fix everything, and everybody could position the plugins and installer exactly where they want them, without them jumping around after every update.

Ruslan-Aleev commented:
>  As for alpha sorting, the sorting would have to be done depending on the language selected for the manager itself.

Sorting will be done by lexicon key, not translation. Make sorting by translation - is a difficult task.

> Why can't we just install a system setting to determine where to put the installer, instead of trying to force everyone to have it in the same place?

In my opinion, the fact that the menu will be on top suits most of the users (based on the discussion in the issue/PR), I didn't think there would be any disputes :)

sebastian-marinescu commented:
> I really don't understand why the consensus is that the installer should always be at the top. Once a site is set up, that is going to be the LEAST USED item in the list. Putting it at the bottom, or else just sorting it like everything else, makes a lot more sense to me.

@rainbowtiger 
I strongly disagree, as it still remains the most important menu-link inside the components-menu (e. g. for maintenance/updates or by the fact that a website is constantly growing and in need of new or swapped extras over time). 
The corresponding issue #11694 has been around *for years*, and the consensus is clear. 

That's why I also feel like @Ruslan-Aleev when it jumps around.
> Personally, I am enraged that the "Installer" jumps in menu :)

And as it is the most important link, putting it as first is a good choice for UX, as also pointed out. 

However, you have valid points in your [comment](https://github.com/modxcms/revolution/pull/14482#issuecomment-660125660) about the different other issues, especially that maybe the *rest* of the menu should be sorted alphanumerically and not by menuindex 👍

rthrash commented:
Top +100. After you have set up a site you should still maintain your extras and keep them up to date. This makes that ever-so-slightly less inconvenient. 

I doubt a discussion about its location would have ever cropped up had it been pinned at the top from the beginning.

SnowCreative commented:
> > As for alpha sorting, the sorting would have to be done depending on the language selected for the manager itself.
> 
> Sorting occurs by lexicon key, not translation.
> 

If plugins were to be sorted alphabetically, they would be sorted by the lexicon values (the translations), not the "keys". No menus are sorted this way now, so this would have to be an added feature. The complexity of this approach is just another good argument for preventing plugin updates from altering the menuindexes at all, so that nothing would jump around after an update.

SnowCreative commented:
> I strongly disagree, as it still remains the most important menu-link inside the components-menu (e. g. for maintenance/updates or by the fact that a website is constantly growing and in need of new or swapped extras over time).

Keeping plugins updated is what is important, not the location of the installer menu item. It you know where that menu item is always going to be, regardless of location, then what's the problem? Adding or swapping plugins over time is a very occasional need. I use SiteDash to do my plugin updating across all sites, so it's not AT ALL important to me to have the installer at the top of the menu in individual sites. An intuitive and useful menu arrangement for my clients is far more important to me. I want to be able to find ALL of my plugins easily, not just the installer. And, I would argue that other plugins in my sites are just as "important" as the installer. 

> However, you have valid points in your [comment](https://github.com/modxcms/revolution/pull/14482#issuecomment-660125660) about the different other issues, especially that maybe the _rest_ of the menu should be sorted alphanumerically and not by menuindex 👍

Actually, alpha sorting is only a second-best option, as far as I'm concerned. Leaving the menuindexes alone is by far the most preferable, because then everybody could set up this menu however they want, and nothing would shift after updates. "Creative Freedom", right? For example, if I have a client who uses a particular plugin every day, I want that to be at the top of the menu, not the installer or some other plugin that comes before it alphabetically.

JoshuaLuckers commented on core/model/modx/modmenu.class.php:
```diff
@@ -101,6 +101,8 @@ public function getSubMenus($start = '') {
         $c->where(array(
             'modMenu.parent' => $start,
         ));
+        // make sure installer is always on top
```
Let's give users a choice and make it optional (simply on/off) in 2.x.

Ruslan-Aleev replied:
If we make a system setting, then it should be enabled by default. It doesn't make much sense, though. Most users favor placing the "Installer" at the top.

SnowCreative replied:
Excellent! Great for in-the-minority folks like me who want it elsewhere.

JoshuaLuckers replied:
And we should fix the issue with the menu indexes in 3.x and look into the other great suggestions!

Jako replied:
Updating an extra resets not only the menuindex. It resets also the parent of menu entries. Both should stay the same.

SnowCreative replied:
It also resets the permissions.

Jako replied:
Ok, that should be moved to a different issue.

sebastian-marinescu replied:
I added a new system-setting `package_installer_at_top` and an if-clause around the installer-sorting

sebastian-marinescu committed:
Update transport.core.system_settings.php

Add `package_installer_at_top` system setting

sebastian-marinescu committed:
Update modmenu.class.php

Use `package_installer_at_top` to determine if installer-menu-entry should be on top

sebastian-marinescu commented:
I realized this PR is still open, so I attempted to add a new system-setting `package_installer_at_top`, which puts the installer-menu-entry at the top, if activated.

Jako commented:
In case you don't know it: https://extras.modx.com/package/menukeeper

sebastian-marinescu commented:
> In case you don't know it: https://extras.modx.com/package/menukeeper

Nice, I didn't! I'm still hoping to get this PR through though 😄

SnowCreative commented:
> In case you don't know it: https://extras.modx.com/package/menukeeper

Looks great!

Jako commented:
The PR misses an english lexicon entry.

sebastian-marinescu committed:
Add EN lexicon-entry for package_installer_at_top

sebastian-marinescu committed:
Add DE lexicon-entry for package_installer_at_top

sebastian-marinescu committed:
Add RO lexicon-entry for package_installer_at_top

sebastian-marinescu commented:
I've added lexicon-entries for the languages I can handle. 

@Jako Can you please check again, if it's okay?

Jako commented:
English should be the only addition. The others are handled by Crowdin.

sebastian-marinescu commented:
Okay, missed that part. So should I delete the other 2 language-entries? Or can it stay there and will be overridden?

Jako commented:
You have to remove the de and ro lexicon files.

sebastian-marinescu committed:
Remove DE entries

sebastian-marinescu committed:
Remove RO entries

sebastian-marinescu commented:
Done! And thank you @Jako, it's my first time committing a PR with system-settings 👍

sebastian-marinescu commented:
Hi @opengeek, can we get this into [v2.8.6](https://github.com/modxcms/revolution/milestone/104)? 

And `port-to-v3` means someone would cherrypick or merge this also into 3.x or do I need to open a separate/new PR for 3.x?

opengeek commented:
> Hi @opengeek, can we get this into [v2.8.6](https://github.com/modxcms/revolution/milestone/104)?
> 
> And `port-to-v3` means someone would cherrypick or merge this also into 3.x or do I need to open a separate/new PR for 3.x?

We can get it into 2.8.6 if we get a consensus on approval (at least one more approving review). And I could attempt to cherry-pick it for 3.x, but it would be preferable if a PR explicitly targeting 3.x was submitted.

sebastian-marinescu commented:
Yay!! 🎉 

Preparing another PR for 3.x now 👍

smg6511 commented on core/lexicon/en/setting.inc.php:
```diff
@@ -501,6 +501,9 @@
 $_lang['setting_new_folder_permissions_desc'] = 'When creating a new folder in the File Manager, the File Manager will attempt to change the folder permissions to those entered in this setting. This may not work on some setups, such as IIS, in which case you will need to manually change the permissions.';
 
+$_lang['setting_package_installer_at_top'] = 'Pin Package-Installer at top';
+$_lang['setting_package_installer_at_top_desc'] = 'If enabled, the package-installer menu-entry inside the Extras-menu will be pinned at the top - regardless of its "menuindex"-attribute. If disabled, the package-installer will be positioned according to its menuindex.';
```
Same as my suggestion for the 3.x version of this PR, tweak to make more concise.
```suggestion
$_lang['setting_package_installer_at_top_desc'] = 'If enabled, the Installer entry will be pinned to the top of the Extras menu. Otherwise it will be positioned according to its menuindex.';
```

smg6511 reviewed (changes requested):
Tested and can approve with requested lexicon change mentioned above (assuming others agree with my tweaks). And, yes I'm one who'd always like to see the Installer at the top, so I do like this feature ;-)

Jako commented:
I like to have some folders as the first entry and because of this I wrote the MenuKeeper extra.

sebastian-marinescu committed:
Update core/lexicon/en/setting.inc.php

Co-authored-by: Jim Graham <info@sparkmediagroup.com>

sebastian-marinescu commented:
Should be good now, right? @opengeek @smg6511

sebastian-marinescu commented:
Thanks @smg6511 👍 

@opengeek 
It seems we got all approvals, can we stil get this into v2.8.6 pretty please?

sebastian-marinescu commented:
And... is something wrong with the cla-bot, or do I have to sign it again? 

![image](https://github.com/modxcms/revolution/assets/3798871/82e90da0-51b8-4b62-be03-a27306f11830)

sebastian-marinescu commented:
@Mark-H @opengeek @JoshuaLuckers @smg6511 

Can one of you, super-pretty please, label it with the [v2.8.6-milestone](https://github.com/modxcms/revolution/milestone/104), so it doesn't slip away?

Also I think the "port-to-v3" label isn't necessary anymore, as the v3-pr went through. 
~And out of curiosity: the label "blocked" is because of the currently not working cla-bot?~
And after the updates, it should also not be labeled "blocked" anymore.

sebastian-marinescu commented:
Thank you very much, Mark! Now I'm calm :))

opengeek commented:
@sebastian-marinescu — I have not yet integrated this PR because @Mark-H and @JoshuaLuckers still have requested changes and did not approve their reviews.

@Mark-H — are you approving this since you added the v2.8.6 label? I'm confused by the comment thread here.

sebastian-marinescu commented:
Thank you for the clarification @opengeek 
IIRC I solved all change requests, but we can wait for Mark and Joshua to confirm 👍

rthrash commented:
@opengeek I believe that @Mark-H is good with the update after all the changes made below his change request and his assigning the previous milestone. @JoshuaLuckers is now outdated, since the toggle was added to the system settings, so should be good to go. Mark, Joshua, please confirm if my review of the reviews is correct! :)
